h2o-algos/src/main/java/hex/tree/sdt/SDT.java

```diff
@@ -0,0 +1,394 @@
+                IntStream.range(0, data.numCols() - 1 /*exclude the last prediction column*/)
+                        .mapToObj(i -> data.vec(i))
+                        // decrease min as the minimum border is always excluded and real min value could be lost
+                        .map(v -> new FeatureLimits(v.min() - EPSILON, v.max()))
+                        .collect(Collectors.toList()));
+    }
+
+
+    private class SDTDriver extends Driver {
+        
+        private void sdtChecks() {
+            if(_train.hasNAs()) {
+                error("_train", "NaNs are not supported yet");
+            }
+            if(_train.hasInfs()) {
+                error("_train", "Infs are not supported");
+            }
+            if(IntStream.range(0, _train.numCols() - 1) // ignore prediction column
+                    .mapToObj(index -> _train.vec(index).isCategorical()).anyMatch(i -> i)) {
+                error("_train", "Categorical features are not supported yet");
+            }
+            if(!_response.isCategorical() ) {
+                error("_response", "Only categorical response is supported");
+            }
+            if(!_response.isBinary()) {
+                error("_response", "Only binary response is supported");
+            }
+        }
+
+        @Override
+        public void computeImpl() {
+            _model = null;
+            try {
+                init(true);
+                sdtChecks();
+                if (error_count() > 0) {
+                    throw H2OModelBuilderIllegalArgumentException.makeFromBuilder(SDT.this);
+                }
+                _rand = RandomUtils.getRNG(_parms._seed);
+                _model = new SDTModel(dest(), _parms,
+                        new SDTModel.SDTOutput(SDT.this));
+                _model.delete_and_lock(_job);
+                buildSDT();
+                LOG.info(_model.toString());
+            } finally {
+                if (_model != null)
+                    _model.unlock(_job);
+            }
+        }
+
+        /**
+         * Build SDT and update infrastructure.
+         */
+        private void buildSDT() {
+            buildSDTIteratively();
+            Log.debug("depth: " + _parms._max_depth + ", nodes count: " + _nodesCount);
+
+            CompressedSDT compressedSDT = new CompressedSDT(_tree);
+
+            _model._output._treeKey = compressedSDT._key;
+            DKV.put(compressedSDT);
+            _job.update(1);
+            _model.update(_job);
+//            System.out.println("Tree: " + compressedSDT.toString());
+            System.out.println("Rules: " + String.join("\n", compressedSDT.getListOfRules()));
+            Log.debug("Tree:");
+            Log.debug(Arrays.deepToString(_tree));
+        }
+
+        /**
+         * Build the tree iteratively starting from the root node.
+         */
+        private void buildSDTIteratively() {
+            _tree = new double[(int) Math.pow(2, _parms._max_depth + 1)][3];
```

Comment:
I tried to run the code but it does not work.  When I hit math_depth=50, I always hit VM out of memory.  Using an array is simple but is probably not the way to go.  Please check GBM or random forest implementation and see how they keep track of their trees and tree nodes.

Reply:
This will be the bottleneck for sure, the maximum array size in Java be 2^31 – 1. IMHO it is enough for the initial implementation since also min_rows parameter comes into the show that will reduce splitting. 

DRF also uses arrays of nodes, but the arrays is resized on demand, so it reflect the min_rows and the type of data itself and do not require the full array at the start but get to it if its needed. The bottle neck of the size is the same though, it just does't directly fail when you give max_depth=50 but it will fail if you provide the dataset large enough to have depth > 31.

Personally, for the initial implementation, I would avoid user to be able have max_depth crazy big and directly fail with: `Requested array size exceeds VM limit`

Secondly, I would check what happen and if cluster correctly fail, when its runs out for memory in the middle of computation. (It is difficult, and even our current tree algorithms doesn't work perfectly)..

The reason we go with the new algorithm is to have nice and clean API and avoid complexity of DTree.